master/internal/configpolicy/task_config_policy_intg_test.go

```diff
@@ -352,3 +363,380 @@ func defaultExperimentConfig() expconf.ExperimentConfigV0 {
+	defer cleanup()
+	db.MustMigrateTestPostgres(t, pgDB, db.MigrationsFromDB)
+
+	user := db.RequireMockUser(t, pgDB)
+	ctx := context.Background()
+	w := createWorkspaceWithUser(ctx, t, user.ID)
+
+	emptyConfig := expconf.ExperimentConfigV0{}
+	defaultConfig := schemas.WithDefaults(emptyConfig)
+
+	wkspDefaultConfig := `{
+	"description": "random description workspace",
+	"resources": {
+		"slots": 5
+	}, 
+	"preemption_timeout": 2000
+}`
+
+	userPartialConfig := `{
+	"description": "random description workspace",
+	"raw_data": { "1" : "data point 1" },
+  	"resources": {
+    "slots": 5,
+    "devices": [
+      {
+        "host_path": "random/path/wksp",
+        "container_path": "random/container/path/wksp",
+		"mode": "slow"
+      }
+    ]
+  },
+  "preemption_timeout": 2000,
+  "bind_mounts": [
+    {
+      "host_path": "random/path",
+      "container_path": "random/container/path/wksp",
+	  "read_only": true,
+	  "propagation": "cluster-wide"
+    }
+  ],
+  "environment": {
+    "environment_variables": {
+      "cpu": [
+        "cpu1:cpuval1, cpu2:cpuval2"
+      ],
+      "proxy_ports": [
+        {
+          "proxy_port": 91950,
+          "proxy_tcp": true
+        }
+      ]
+    }
+  },
+  "hyperparameters": {
+    "categorical": {
+      "vals": [
+        27, 30
+      ]
+    }
+  },
+  "log_polcies": [
+    {
+      "pattern": "nonrepeat",
+      "action": {
+        "cancel_retries": 10
+      }
+    }
+  ]
+}
+`
+
+	wkspPartialConfig := `{
+  "raw_data": { "2" : "data point 2" },
+  "bind_mounts": [
+    {
+      "host_path": "random/path",
+      "container_path": "random/container/path",
+	  "read_only": true,
+	  "propagation": "cluster-wide"
+    }
+  ],
+  "environment": {
+    "environment_variables": {
+      "cpu": [
+        "cpu1:cpuval1"
+      ],
+      "proxy_ports": [
+        {
+          "proxy_port": 9195,
+          "proxy_tcp": true
+        }
+      ]
+    }
+  },
+  "hyperparameters": {
+    "categorical": {
+      "vals": [
+        27
+      ]
+    }
+  },
+  "log_polcies": [
+    {
+      "pattern": "repeat",
+      "action": {
+        "cancel_retries": 1
+      }
+    }
+  ]
+}`
+
+	mergedWkspConfig := `
+{
+  "raw_data": { 
+  		"1" : "data point 1",
+ 		"2" : "data point 2"
+	},
+  "description": "random description workspace",
+  "resources": {
+    "slots": 5,
+    "devices": [
+      {
+        "host_path": "random/path",
+        "container_path": "random/container/path",
+        "mode": "fast"
+      },
+      {
+        "host_path": "random/path/wksp",
+        "container_path": "random/container/path/wksp",
+        "mode": "slow"
+      }
+    ]
+  },
+  "preemption_timeout": 2000,
+  "bind_mounts": [
+    {
+      "host_path": "random/path",
+      "container_path": "random/container/path/wksp",
+      "read_only": true,
+      "propagation": "cluster-wide"
+    }
+  ],
+  "environment": {
+    "environment_variables": {
+      "cpu": [
+        "cpu1:cpuval1, cpu2:cpuval2"
```

Comment:
I'm not able to follow what's being merged in the test cases and what the expected output is. I want to double-check that we're testing a merge, not just replacement. We want a case like 

global env variables:
```
  "environment": {
    "environment_variables": {
      "cpu": [
        "cpu1:cpuval1, cpu2:cpuval2"
```

workspace env variables:
```
  "environment": {
    "environment_variables": {
      "cpu": [
        "cpu3:cpuval3"
 ```
 
 user env variables:
 ```  "environment": {
    "environment_variables": {
      "cpu": [
        "cpu4:cpuval4"
```

merged env variables (global + user but not workspace):
```
  "environment": {
    "environment_variables": {
      "cpu": [
        "cpu1:cpuval1, cpu2:cpuval2, cpu4:cpuval4"
```

Reply:
The test for merged slices and maps is a bit confusing, I'll move it into its own test function to clarify. 
Also, just to clarify how I tested maps and slice merge logic:

Since maps being merged during the unmarshal process is part of the Go standard library `json` and `yaml` packages, I only tested one map field in the maps and slices merge test case.
Since the logic behind whether a slice gets merged varies from slice to slice, I tested all of the slices whose wrapper objects implement our `Mergable` pseudointerface (and documented those that didn't). 

The differing JSON fields for all the aforementioned slices (and a singular map) are provided in both `userPartialInvariantConfig` and `wkspPartialInvariantConfig` as differing fields, with `mergedWkspInvariantConfig` being the combination of both (non map/slice fields with normal override behavior, and slices and maps expected to be merged). So, for each field, `userPartialInvariantConfig` has 1 val, `wkspPartialInvariantConfig` has 1 val, and `mergedWkspInvariantConfig` has 2.

This test definitely belongs separate from the rest of the test-driven cases as it has a complex and unique setup. I'll move it to its own function and refactor it so that its function is easier to understand